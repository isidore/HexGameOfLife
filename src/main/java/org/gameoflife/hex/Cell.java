package org.gameoflife.hex;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

public class Cell {
    private int x;
    private int y;

    public Cell(int x, int y) {

        this.x = x;
        this.y = y;
    }

    public static Cell create(int x, int y) {
        return new Cell(x,y);
    }

    public static List<Cell> getLevelOneNeighbours(int x, int y) {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x - 1, y-1));
        cells.add(new Cell(x - 1, y+1));
        cells.add(new Cell(x , y-2));
        cells.add(new Cell(x , y+2));
        cells.add(new Cell(x +1, y-1));
        cells.add(new Cell(x + 1, y+1));

        return cells;
    }

    public static List<Cell> getLevelTwoNeighbours(int x, int y) {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x-2, y));
        cells.add(new Cell(x -1, y+3));
        cells.add(new Cell(x - 1, y-3));
        cells.add(new Cell(x + 1, y+3));
        cells.add(new Cell(x +1, y-3));
        cells.add(new Cell(x+2 , y));

        return cells;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Cell cell = (Cell) o;
        return x == cell.x && y == cell.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    public List<Cell> getLevelOneNeighbours() {
        return getLevelOneNeighbours(x,y);
    }

    public List<Cell> getLevelTwoNeighbours() {
        return getLevelTwoNeighbours(x, y);
    }
}
