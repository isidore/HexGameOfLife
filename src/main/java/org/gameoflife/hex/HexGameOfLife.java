package org.gameoflife.hex;

import org.apache.commons.lang.math.DoubleRange;
import org.lambda.query.Queryable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class HexGameOfLife implements GameOfLife {

    final Board board;

    public HexGameOfLife() {
        this(new ArrayList<>());
    }

    public HexGameOfLife(List<Cell> liveCells) {
        this.board = new Board();
        this.board.setLiveCells(liveCells);
    }

    public static boolean survivesToNextTurn(double sum, boolean alive) {
        DoubleRange survivable = new DoubleRange(2, 3.3);
        DoubleRange growth = new DoubleRange(2.3, 2.9);

        boolean survives = survivable.containsDouble(sum) && alive;
        boolean born = growth.containsDouble(sum);
        return survives || born;
    }

    @Override
    public List<Cell> getLiveCells() {
        return board.getLiveCells();
    }

    @Override
    public HexGameOfLife advanceTurn() {
        Queryable<Cell> nextLivingCells = getLiveCellsAndNeighbours()
                .where(c -> HexGameOfLife.survivesToNextTurn(getNeighbourScore(board, c), board.isAlive(c)));

        return new HexGameOfLife(nextLivingCells);
    }

    private static double getNeighbourScore(Board board, Cell cell) {
        double score = 0.0;
        score += getScore(board, cell.getLevelOneNeighbours(), 1);
        score += getScore(board, cell.getLevelTwoNeighbours(), 0.3);
        return score;
    }

    private static double getScore(Board board, List<Cell> neighbours, double weight) {
        double score = 0;

        for (Cell cell : neighbours) {
            if (board.isAlive(cell)) {
                score += weight;
            }
        }

        return score;
    }

    @Override
    public void setAlive(int x, int y) {
        this.board.setAlive(x, y);
    }

    private Queryable<Cell> getLiveCellsAndNeighbours() {
        Set<Cell> liveCellsAndNeighbours = new HashSet<>(board.getLiveCells());

        for (Cell cell : board.getLiveCells()) {
            liveCellsAndNeighbours.addAll(cell.getAllNeighbours());
        }

        return Queryable.as(new ArrayList<>(liveCellsAndNeighbours));
    }

    @Override
    public String toString() {
        return HexPrinter.print(this);
    }

    @Override
    public boolean isAlive(Cell cell) {
        return board.isAlive(cell);
    }

}
