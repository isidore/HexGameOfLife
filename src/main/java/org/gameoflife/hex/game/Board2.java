package org.gameoflife.hex.game;


import java.util.List;

public class Board2 {
    private List<Cell> liveCells;

    public List<Cell> getLiveCells() {
        return liveCells;
    }

    void setLiveCells(List<Cell> liveCells) {
        this.liveCells = liveCells;
    }

    void setAlive(int x, int y) {
        liveCells.add(new Cell(x, y));
    }

    boolean isAlive(Cell cell) {
        return liveCells.contains(cell);
    }

}
