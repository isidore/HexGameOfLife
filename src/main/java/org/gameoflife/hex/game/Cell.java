package org.gameoflife.hex.game;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Cell {
    private final int x;
    private final int y;

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Cell(int x, int y) {

        this.x = x;
        this.y = y;
    }

    List<Cell> getAllNeighbours() {
        ArrayList<Cell> neighbours = new ArrayList<>();
        neighbours.addAll(getLevelOneNeighbours());
        neighbours.addAll(getLevelTwoNeighbours());
        return neighbours;
    }

    public List<Cell> getLevelOneNeighbours() {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x - 1, y - 1));
        cells.add(new Cell(x + 1, y - 1));
        cells.add(new Cell(x - 2, y));
        cells.add(new Cell(x + 2, y));
        cells.add(new Cell(x - 1, y + 1));
        cells.add(new Cell(x + 1, y + 1));

        return cells;
    }

    public List<Cell> getLevelTwoNeighbours() {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x, y - 2));
        cells.add(new Cell(x + 3, y - 1));
        cells.add(new Cell(x - 3, y - 1));
        cells.add(new Cell(x + 3, y + 1));
        cells.add(new Cell(x - 3, y + 1));
        cells.add(new Cell(x, y + 2));

        return cells;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Cell cell = (Cell) o;
        return x == cell.x && y == cell.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return "(" + x + "," + y + ")";
    }
}
