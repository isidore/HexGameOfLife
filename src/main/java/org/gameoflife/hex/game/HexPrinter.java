package org.gameoflife.hex.game;

import dk.ilios.asciihexgrid.AsciiBoard;
import dk.ilios.asciihexgrid.printers.SmallPointyAsciiHexPrinter;

public class HexPrinter {
    public static String print(HexGameOfLife game) {
        AsciiBoard asciiBoard = new AsciiBoard(0, 100, 0, 100, new SmallPointyAsciiHexPrinter());
        int size = 10;
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                if (HexGameOfLife.isValidCoordinates(new Coordinates(x, y))) {
                    if (game.isAlive(new Cell(x, y))) {
                        asciiBoard.printHex("X", "X", 'X', translateX(x, y), translateY(x, y));
                    } else {
                        asciiBoard.printHex(x + "," + y, "", ' ', translateX(x, y), translateY(x, y));
                    }
                }
            }
        }
        return asciiBoard.prettPrint(true);
    }

    private static int translateY(int x, int y) {
        return y;
    }

    private static int translateX(int x, int y) {
        int emptyspaces = (x + 1) / 2;
        return x - emptyspaces;
    }
}
