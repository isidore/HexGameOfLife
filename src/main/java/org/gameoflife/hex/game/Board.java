package org.gameoflife.hex.game;


import org.gameoflife.hex.Cell;

import java.util.List;

public class Board {
    private List<Cell> liveCells;

    List<Cell> getLiveCells() {
        return liveCells;
    }

    void setLiveCells(List<Cell> liveCells) {
        this.liveCells = liveCells;
    }

    void setAlive(int x, int y) {
        liveCells.add(new Cell(x, y));
    }

    boolean isAlive(Cell cell) {
        return liveCells.contains(cell);
    }

}
