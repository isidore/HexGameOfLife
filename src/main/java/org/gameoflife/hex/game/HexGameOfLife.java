package org.gameoflife.hex.game;

import com.spun.util.FormattedException;
import org.apache.commons.lang.math.DoubleRange;
import org.gameoflife.hex.Coordinates;
import org.gameoflife.hex.GameOfLife;
import org.gameoflife.hex.HexPrinter;
import org.lambda.query.Queryable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class HexGameOfLife implements GameOfLife {

    final List<Cell> board;

    public HexGameOfLife() {
        this(new ArrayList<>());
    }

    public HexGameOfLife(List<Cell> liveCells) {
        this.board = liveCells;
    }

    public static boolean survivesToNextTurn(double sum, boolean alive) {
        DoubleRange survivable = new DoubleRange(2, 3.3);
        DoubleRange growth = new DoubleRange(2.3, 2.9);

        boolean survives = survivable.containsDouble(sum) && alive;
        boolean born = growth.containsDouble(sum);
        return survives || born;
    }

    public HexGameOfLife advanceTurn() {
        Queryable<Cell> nextLivingCells = getLiveCellsAndNeighbours()
                .where(c -> HexGameOfLife.survivesToNextTurn(getNeighbourScore(board, c), board.contains(c)));

        return new HexGameOfLife(nextLivingCells);
    }

    private static double getNeighbourScore(List<Cell> board, Cell cell) {
        double score = 0.0;
        score += getScore(board, cell.getLevelOneNeighbours(), 1);
        score += getScore(board, cell.getLevelTwoNeighbours(), 0.3);
        return score;
    }

    private static double getScore(List<Cell> board, List<Cell> neighbours, double weight) {
        double score = 0;

        for (Cell cell : neighbours) {
            if (board.contains(cell)) {
                score += weight;
            }
        }

        return score;
    }

    public void setAlive(int x, int y) {
        if (!GameOfLife.isValidCoordinates(new Coordinates(x, y))) {
            throw new FormattedException("Invalid Location for (%s, %s)", x, y);
        }

        this.board.add(new Cell(x, y));
    }

    private Queryable<Cell> getLiveCellsAndNeighbours() {
        Set<Cell> liveCellsAndNeighbours = new HashSet<>(board);

        for (Cell cell : board) {
            liveCellsAndNeighbours.addAll(cell.getAllNeighbours());
        }

        return Queryable.as(new ArrayList<>(liveCellsAndNeighbours));
    }

    public String toString() {
        return HexPrinter.print(this);
    }

    public boolean isAlive(Cell cell) {
        return board.contains(cell);
    }

    public List<Cell> getLiveCells() {
        return board;
    }
}
