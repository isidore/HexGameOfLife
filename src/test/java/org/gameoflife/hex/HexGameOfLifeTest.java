package org.gameoflife.hex;

import org.approvaltests.Approvals;
import org.gameoflife.hex.game.Cell;
import org.gameoflife.hex.game.HexGameOfLife;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

public class HexGameOfLifeTest {


    @Test
    static void testExceptionIsThrownForInvalidCoordinate() {
        Approvals.verifyException(() ->
        {
            HexGameOfLife hexGameOfLife = new HexGameOfLife();
            hexGameOfLife.setAlive(1, 2);
        });
    }

    @Test
    void testZeroDies() {
        String timeline = "";
        HexGameOfLife hexGameOfLife = new HexGameOfLife();
        hexGameOfLife.setAlive(4, 4);
        timeline += hexGameOfLife;
        HexGameOfLife hexGameOfLife2 = hexGameOfLife.advanceTurn();
        timeline += "\n NEXT TURN \n";
        timeline += hexGameOfLife2;
        Approvals.verify(timeline);
    }

    @Test
    void testTwoFirstLevelNeighbours() {
        String timeline = "";
        HexGameOfLife hexGameOfLife = new HexGameOfLife();
        hexGameOfLife.setAlive(4, 4);
        hexGameOfLife.setAlive(3, 5);
        hexGameOfLife.setAlive(4, 6);
        hexGameOfLife.setAlive(5, 5);
        timeline += HexPrinter.print(hexGameOfLife);
        HexGameOfLife hexGameOfLife2 = hexGameOfLife.advanceTurn();
        timeline += "\n NEXT TURN \n";
        timeline += HexPrinter.print(hexGameOfLife2);
        Approvals.verify(timeline);
    }

    @Test
    void testEverything() {
        String fullTimeline = "";

        for (int l1 = 0; l1 <= 6; l1++) {
            for (int l2 = 0; l2 <= 6; l2++) {
                String timeline = advanceBoardWithNeighbours(l1, l2);
                fullTimeline += timeline + "\n\n\n\n";
            }
        }
        Approvals.verify(fullTimeline);
    }

    private String advanceBoardWithNeighbours(int numberOfLevelOneNeighbours, int numberOfLevelTwoNeighbours) {
        String title = String.format("Neighbours(Level 1 , Level 2) = (%s, %s) => \n", numberOfLevelOneNeighbours, numberOfLevelTwoNeighbours);
        HexGameOfLife hexGameOfLife = createGameWithNeighbours(numberOfLevelOneNeighbours, numberOfLevelTwoNeighbours, true);
        return printAdvanceBoard(title, hexGameOfLife);
    }

    private String printAdvanceBoard(String title, HexGameOfLife hexGameOfLife) {
        String timeline = "";
        timeline += title;


        timeline += HexPrinter.print(hexGameOfLife);
        timeline += "\n NEXT TURN \n";
        HexGameOfLife postTurnBoard = hexGameOfLife.advanceTurn();
        timeline += HexPrinter.print(postTurnBoard);
        return timeline;
    }

    @Test
    void testCentreComesAlive() {
        HexGameOfLife boardWithNeighbours = createGameWithNeighbours(2, 3, false);
        Approvals.verify(printAdvanceBoard("Centre (4,4) comes alive\n", boardWithNeighbours));
    }

    public static HexGameOfLife createGameWithNeighbours(int numberOfLevelOneNeighbours, int numberOfLevelTwoNeighbours, boolean includeCentre) {
        Cell centre = new Cell(4, 4);
        ArrayList<Cell> cells = new ArrayList<>();

        if (includeCentre) {
            cells.add(centre);
        }

        List<Cell> level1 = centre.getLevelOneNeighbours().subList(0, numberOfLevelOneNeighbours);
        cells.addAll(level1);

        List<Cell> level2 = centre.getLevelTwoNeighbours().subList(0, numberOfLevelTwoNeighbours);
        cells.addAll(level2);

        return new HexGameOfLife(cells);
    }
}
