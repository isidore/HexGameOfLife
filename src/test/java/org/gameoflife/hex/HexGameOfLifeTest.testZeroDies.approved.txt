| = = = = = = = = = = = = = = = = = = = = = = = |
|    / \     / \     / \     / \     / \        |
|  /     \ /     \ /     \ /     \ /     \      |
| |  0,0  |  2,0  |  4,0  |  6,0  |  8,0  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,1  |  3,1  |  5,1  |  7,1  |  9,1  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,2  |  2,2  |  4,2  |  6,2  |  8,2  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,3  |  3,3  |  5,3  |  7,3  |  9,3  | |
|     |       |       |       |       |       | |
|    / \     / \     /X\     / \     / \     /  |
|  /     \ /     \ /X X X\ /     \ /     \ /    |
| |  0,4  |  2,4  |X  X  X|  6,4  |  8,4  |     |
| |       |       |X  X  X|       |       |     |
|  \     / \     / \X X X/ \     / \     / \    |
|    \ /     \ /     \X/     \ /     \ /     \  |
|     |  1,5  |  3,5  |  5,5  |  7,5  |  9,5  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,6  |  2,6  |  4,6  |  6,6  |  8,6  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,7  |  3,7  |  5,7  |  7,7  |  9,7  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,8  |  2,8  |  4,8  |  6,8  |  8,8  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,9  |  3,9  |  5,9  |  7,9  |  9,9  | |
|     |       |       |       |       |       | |
|      \     / \     / \     / \     / \     /  |
|        \ /     \ /     \ /     \ /     \ /    |
| = = = = = = = = = = = = = = = = = = = = = = = |

 NEXT TURN 
| = = = = = = = = = = = = = = = = = = = = = = = |
|    / \     / \     / \     / \     / \        |
|  /     \ /     \ /     \ /     \ /     \      |
| |  0,0  |  2,0  |  4,0  |  6,0  |  8,0  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,1  |  3,1  |  5,1  |  7,1  |  9,1  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,2  |  2,2  |  4,2  |  6,2  |  8,2  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,3  |  3,3  |  5,3  |  7,3  |  9,3  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,4  |  2,4  |  4,4  |  6,4  |  8,4  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,5  |  3,5  |  5,5  |  7,5  |  9,5  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,6  |  2,6  |  4,6  |  6,6  |  8,6  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,7  |  3,7  |  5,7  |  7,7  |  9,7  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,8  |  2,8  |  4,8  |  6,8  |  8,8  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,9  |  3,9  |  5,9  |  7,9  |  9,9  | |
|     |       |       |       |       |       | |
|      \     / \     / \     / \     / \     /  |
|        \ /     \ /     \ /     \ /     \ /    |
| = = = = = = = = = = = = = = = = = = = = = = = |
